FEAT(client): Add setting to always show user in Talking UI

jvbsl opened the pull request:
FEAT(client): Add setting to always show user in Talking UI

Allow talking ui user timeout to be -1(meaning that user should never timeout and therefore be always displayed)

Change user display state dependant on settings on startup and settings change

Prevent user display timeout if setting is set to -1


### Checks

- [x] My commits follow the [commit guidelines](https://github.com/mumble-voip/mumble/blob/master/COMMIT_GUIDELINES.md)

I'm not sure if I did everything as expected, but for small servers I think it is nice to always see the users.

Hartmnt commented on src/mumble/LookConfig.ui:
```diff
@@ -636,6 +636,9 @@
          <string>A user that is silent for the given amount of seconds will be removed from the Talkin UI.</string>
         </property>
+        <property name="minimum">
+         <number>-1</number>
```
If I understand the code correctly, ``-1`` will be displayed when your new option takes effect, right? I think that is not very nice from a user perspective.
It would be better, if a string was displayed like "Disabled" or something similar. Alternatively, you could create a check-box to enable/disable this feature and make the slider inactive, if it is disabled.

jvbsl replied:
Yes, you understood it perfectly.
I thought about using a checkbox as well, but then there are some UI things to consider and I'm not really good at that.
E.g. if I add a checkbox to all the checkboxes in the UI above, it will influence the slider which is much further down in the UI and there is no visible indication that they belong together.

If I instead add the checkbox directly to the option it looks strange to have a checkbox out of nowhere, and the question is how indentation should be done:
![grafik](https://github.com/mumble-voip/mumble/assets/13928225/8337b8f0-672d-4724-a0fb-711ee79e6982)

Hartmnt replied:
Maybe place a checkbox to the right of a label that says "Remove silent user". And add the number box to a new row with the label "after...". Then also add the Qt suffix "seconds" to the number box.

But disregarding what we choose here: everything is better than just "-1"

jvbsl replied:
We could do it inside a groupbox but that would destroy the table like layout:
![grafik](https://github.com/mumble-voip/mumble/assets/13928225/62efe5e3-5c80-491b-b797-6eecfc08ef2a)

I could do it with two separate rows, and indent the "after" label manually, but I do not think it is that nice to have one and the same thing span two rows. Any idea?

Hartmnt replied:
What if you place the checkbox to the right of the label instead of to the left. So the checkbox is horizontally aligned with the number boxes? Maybe the "Audio Input" "Misc" section inspires you. There it is also handled slightly differently.

jvbsl replied:
I can't really place it on the right using a groupbox(and tbh I would personally not like the look of that).
Instead I just did it analogous to the "Abbreviate channel names":
![grafik](https://github.com/mumble-voip/mumble/assets/13928225/766572bc-8149-4ca9-ab47-d941bcbbafa5)

Do you think that would be acceptable?

Hartmnt replied:
Yep that is neat. :+1:

Hartmnt reviewed (commented):
I would always recommend first creating an Issue here on GitHub before working on a new feature so it can be discussed by everyone. Otherwise, you might risk that the feature is considered out of scope and rejected. For the record, I personally think this feature is probably useful and worth implementing.

jvbsl commented:
First of all thank you very much for your fast reply.
Should I resolve each of your remarks in a separate commit, or in a single commit?
Or will it get squashed anyway?

jvbsl commented on src/mumble/TalkingUI.cpp:
```diff
@@ -395,9 +395,12 @@ TalkingUIUser *TalkingUI::findOrAddUser(const ClientUser *user) {
 		// * 1000 as the setting is in seconds whereas the timer expects milliseconds
-		userEntry->setLifeTime(Global::get().s.iTalkingUI_SilentUserLifeTime * 1000);
+		int silentUserLifeTime = Global::get().s.iTalkingUI_SilentUserLifeTime;
+		if (silentUserLifeTime != -1)
```
Good to know, is there a document where all the guidelines are written down? I didn't see it in the Coding Guidelines.
Just fyi we could add "InsertBraces: true" to the clang-format file, but it is only supported with clang-format >= 15 if you are comfortable with having such a "high" version requirement.

Hartmnt replied:
No, haha. That got me too when I submitted my first pull request. The Mumble style requirements are currently only living in the head of @Krzmbrzl . Maybe this can be written down somewhere at some point :)

jvbsl commented on src/mumble/TalkingUI.cpp:
```diff
@@ -395,9 +395,12 @@ TalkingUIUser *TalkingUI::findOrAddUser(const ClientUser *user) {
+			userEntry->setLifeTime(silentUserLifeTime * 1000);
 
-		userEntry->restrictLifetime(!isSelf || !Global::get().s.bTalkingUI_LocalUserStaysVisible);
+		userEntry->restrictLifetime(silentUserLifeTime != -1
```
Do you mean to change the restrictLifetime call to take in two boolean values instead, or just use a temporary variable instead of the `!= -1` check? Or do you mean to go one step further and have a separate global boolean variable?

Hartmnt replied:
I am thinking of using one or two temporary bools to make the call spanning line 402 and 403 easier to read

Hartmnt commented:
> First of all thank you very much for your fast reply. Should I resolve each of your remarks in a separate commit, or in a single commit? Or will it get squashed anyway?

Squash everything into your existing commit. The simplest way to do this is to always use ``git commit --amend`` now and force-pushing going forward.
If you have already added new commits, you can use ``git rebase HEAD~N --interactive`` where N is the number of commits you want to edit. Then change the action from ``pick`` to ``squash`` for every commit except the first. (force push is also required here then)

jvbsl commented:
Now just one more thing :D
If I used InsertBrackets: true locally to format the source code files, but it reformatted code that actually was never touched by me, is it fine to keep it in the commit, or not? Normally I would not keep it, but then again commiting guidelines mentioned to always execute clang-format before commiting(but my clang-format contained the new mumble guidelines in respect to single line ifs - and I hope it also extends to single line for loops?)

Hartmnt commented:
> Now just one more thing :D If I used InsertBrackets: true locally to format the source code files, but it reformatted code that actually was never touched by me, is it fine to keep it in the commit, or not? Normally I would not keep it, but then again commiting guidelines mentioned to always execute clang-format before commiting(but my clang-format contained the new mumble guidelines in respect to single line ifs - and I hope it also extends to single line for loops?)

No, please do not commit code formatting changes in parts of the code that you did not explicitly touch. Yes, I think single line loops would need the brackets, too.

jvbsl commented:
Alright, sorry two more questions^^
I did not mean that I did not touch the code files, the changes would be in the files that I actually edited anyway, still only format the lines I changed, or should I format the files that I changed?

And what about translations, I can't find the location where translations are for the UI settings and should I even do that in this PR?

P.S.: Next PR should be simpler then :D

Hartmnt commented:
> I did not mean that I did not touch the code files, the changes would be in the files that I actually edited anyway, still only format the lines I changed, or should I format the files that I changed?

All files are already formatted with clang-format-10 with the default configuration. The problem is that clang-format-10 allows for both: single line ``if``s with and without brackets in its default configuration. It will not change one into the other automatically. @Krzmbrzl decided that we only accept single line ``if``s and probably also ``for``s with brackets. You should be able to clang-format-10 all files and the ``if`` brackets wont change. However, if you change the configuration to force it to use brackets, you might change code that has not been touched by this merge request.

> And what about translations, I can't find the location where translations are for the UI settings and should I even do that in this PR?

Once everything else is done, you can run the [updatetranslations.py](https://github.com/mumble-voip/mumble/blob/master/scripts/updatetranslations.py) script from the repository root. That will automatically create an additional translation commit using your git settings. That translation commit is indeed supposed to be a second commit next to your code change. So in the end your merge request will contain two commits. (Note: Obviously, the translation commit is only necessary, if you change translatable strings)

jvbsl commented:
Should I actually try translating the things I added to the different languages using translation tools?
Or only translate to the language I know?
Or should that all happen in a different PR?

Better translated badly or rather not at all?

Hartmnt commented:
> Should I actually try translating the things I added to the different languages using translation tools? Or only translate to the language I know? Or should that all happen in a different PR?
> 
> Better translated badly or rather not at all?

Translations happen via weblate, an external service. You can contribute there, if and only if you like. No need to translate in merge requests.

https://github.com/mumble-voip/mumble#translating

Hartmnt commented on src/mumble/Settings.cpp:
```diff
@@ -1005,6 +1005,7 @@ void Settings::legacyLoad(const QString &path) {
 	// TalkingUI
 	LOAD(qpTalkingUI_Position, "ui/talkingUIPosition");
 	LOAD(bShowTalkingUI, "ui/showTalkingUI");
+	LOAD(bTalkingUI_UsersAlwaysVisible, "ui/talkingUI_UsersAlwaysVisible");
```
You can remove this line as this is the loading routine for the legacy settings (in INI format). Mumble has switched to JSON and since your new option by definition not be included in legacy config files, you do not have to load it here.

jvbsl replied:
Oh alright, I was wondering why it was loaded at two locations, but blessed are the people who can read "legacyLoad". Perhaps I should actually use an IDE for mumble^^

Hartmnt commented on src/mumble/TalkingUI.cpp:
```diff
@@ -397,7 +397,10 @@ TalkingUIUser *TalkingUI::findOrAddUser(const ClientUser *user) {
+		bool isLocalUserAndLocalUserAlwaysVisible = isSelf && Global::get().s.bTalkingUI_LocalUserStaysVisible;
+		bool usersAlwaysVisible                   = Global::get().s.bTalkingUI_UsersAlwaysVisible;
+		bool isUserAlwaysVisible                  = usersAlwaysVisible || isLocalUserAndLocalUserAlwaysVisible;
+		userEntry->restrictLifetime(!isUserAlwaysVisible);
```
nice :+1:

jvbsl commented:
I'm so sorry I missed the minimum value in the LookConfig.ui it was still set to -1

Hartmnt commented on src/mumble/TalkingUIEntry.cpp:
```diff
@@ -270,7 +270,7 @@ void TalkingUIUser::restrictLifetime(bool restrict) {
 		// Start timer
 		m_timer.start();
-	} else if (!restrict && !m_timer.isActive()) {
+	} else if (!restrict && m_timer.isActive()) {
```
Ha, you found a bug in code that was never reached!

Hartmnt commented on src/mumble/Settings.h:
```diff
@@ -369,6 +369,7 @@ struct Settings {
 
 	QPoint qpTalkingUI_Position              = UNSPECIFIED_POSITION;
 	bool bShowTalkingUI                      = false;
+	bool bTalkingUI_UsersAlwaysVisible       = false;
```
Ok, one last thing (sorry). Could you please remove the type prefix "b" here? This is one of the new Mumble style requirements. Should have noticed earlier, but I am usually of the other side of code reviews ;)

jvbsl replied:
No problem I fixed it now ;)

Hartmnt reviewed (approved):
Compiled and tested. Works as advertised.

Krzmbrzl reviewed (approved):
Conceptually the little fix in `TalkingUIEntry` should have been its own commit but given this is so minor that nobody even noticed this before, I'm fine with leaving things as they are.

Thank you very much for your work and sorry for having taken so extremely long before reviewing this :pray:

Krzmbrzl commented:
@Hartmnt for the record: excellent code review that you did here :+1:
